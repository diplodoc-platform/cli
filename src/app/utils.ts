import {TextSizes, Theme} from '@doc-tools/components';

const DEFAULT_USER_SETTINGS = {
    theme: Theme.Light,
    textSize: TextSizes.m,
    showMiniToc: true,
    wideFormat: true,
    fullScreen: false,
};

export function getDocSettings() {
    const {
        theme: defaultTheme,
        textSize: defaultTextSize,
        showMiniToc: defaultShowMiniToc,
        wideFormat: defaultWideFormat,
        fullScreen: defaultFullScreen,
    } = DEFAULT_USER_SETTINGS;

    const theme = getSetting('theme') as Theme || defaultTheme;
    const textSize = getSetting('textSize') as TextSizes || defaultTextSize;
    const showMiniToc = getSetting('showMiniToc') || defaultShowMiniToc;
    const wideFormat = getSetting('wideFormat') || defaultWideFormat;
    const fullScreen = getSetting('fullScreen') || defaultFullScreen;

    return {
        theme,
        textSize,
        showMiniToc: strToBoolean(showMiniToc),
        wideFormat: strToBoolean(wideFormat),
        fullScreen: strToBoolean(fullScreen),
    };
}

export function getSetting(name: string) {
    return sessionStorage.getItem(name);
}

export function saveSetting<T>(name: string, value: T) {
    sessionStorage.setItem(name, String(value));
}

export function strToBoolean(str: any) {
    if (typeof str === 'boolean') {
        return str;
    }

    return str ? str === 'true' : false;
}

export function withSavingSetting<T>(settingName: string, onChange: (value: T) => void) {
    return (value: T) => {
        saveSetting<T>(settingName, value);

        onChange(value);
    };
}

export function updateRootClassName(theme: Theme, isMobile = false) {
    const themeClassName = theme === 'light' ? 'yc-root_theme_light' : 'yc-root_theme_dark';
    const mobileClassName = isMobile ? 'mobile' : '';

    document.body.className = `yc-root ${themeClassName} ${mobileClassName}`;
}
