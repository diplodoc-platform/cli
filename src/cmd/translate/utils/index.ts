import {ok} from 'assert';
import {basename, dirname, extname, resolve} from 'path';
import {readFileSync} from 'node:fs';
import glob from 'glob';

export {dumpFile, loadFile, resolveSchemas} from './fs';
export {extract, compose} from './translate';
export {TranslateError, LimitExceed, RequestError, AuthError} from './errors';

type TranslateArgs = {
    input: string;
    output?: string;
    source?: string;
    sourceLanguage?: string;
    sourceLanguageLocale?: string;
    target?: string;
    targetLanguage?: string;
    targetLanguageLocale?: string;
    auth?: string;
    folder?: string;
    glossary?: string;
    include?: string[] | string;
    exclude?: string[] | string;
    dryRun?: boolean;
    useSource?: boolean;
} & {
    [prop: string]: any;
};

export type TranslateParams = {
    input: string;
    output: string;
    source: [string, string];
    targets: [string, string][];
    auth?: string;
    folder?: string;
    glossary?: string;
    files: string[];
    dryRun: boolean;
    useSource: boolean;
};

export function normalizeParams(params: TranslateArgs, exts = EXTS): TranslateParams {
    const source = normalizeLocale(params, 'source')[0];
    const targets = normalizeLocale(params, 'target');
    const {input, files} = normalizeInput(params, source[0], exts);
    const {output, auth, folder, dryRun, useSource} = params;

    ok(input, 'Required param input is not configured');

    return {
        input,
        output: output || input,
        auth,
        folder,
        source,
        targets,
        files: [...new Set(files)],
        dryRun: Boolean(dryRun),
        useSource: Boolean(useSource),
    };
}

function normalizeLocale(params: TranslateArgs, scope: 'source' | 'target'): [string, string][] {
    const _scopeLanguage = params[scope + 'Language'];
    const _scopeLanguageLocale = params[scope + 'LanguageLocale'];
    const _scope = params[scope] || _scopeLanguageLocale || _scopeLanguage;

    if (!_scope) {
        return [['', '']];
    }

    if (Array.isArray(_scope)) {
        return _scope.map((_scope) => _scope.split('-'));
    }

    return [_scope.split('-') as [string, string]];
}

const EXTS = ['.md', '.yaml', '.json'];

function normalizeInput(params: TranslateArgs, language: string, exts: string[]) {
    let {input, include = [], exclude = []} = params;

    include = ([] as string[]).concat(include);
    exclude = ([] as string[]).concat(exclude);

    let files: string[] | null = null;
    if (extname(input) === '.list') {
        const list = readFileSync(input, 'utf8').split('\n');
        input = dirname(input);
        files = list.map((file) => {
            const absPath = resolve(input, file);

            if (!absPath.startsWith(input)) {
                throw new Error(`Insecure access to file out of project scope. (file: ${absPath})`);
            }

            if (!exts.includes(extname(file))) {
                throw new Error(`Unhandles file extension. (file: ${absPath})`);
            }

            return absPath;
        });
    } else {
        if (exts.includes(extname(input))) {
            files = [basename(input)];
            input = dirname(input);
        }

        if (!include.length) {
            include.push('...');
        }

        include = include.reduce((acc, item) => {
            if (item === '...') {
                acc.push(...exts.map((ext) => (language || '.') + '/**/*' + ext));
            } else {
                acc.push(item);
            }

            return acc;
        }, [] as string[]);

        files =
            files ||
            ([] as string[]).concat(
                ...include.map((match) =>
                    glob.sync(match, {
                        cwd: input,
                        ignore: exclude,
                        nodir: true,
                    }),
                ),
            );
    }

    return {input, files};
}

export class Defer<T = string> {
    resolve!: (text: T) => void;

    reject!: (error: any) => void;

    promise: Promise<T>;

    constructor() {
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
    }
}

export function bytes(texts: string[]) {
    return texts.reduce((sum, text) => sum + text.length, 0);
}
