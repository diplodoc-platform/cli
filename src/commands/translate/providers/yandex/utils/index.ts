export {LimitExceed, RequestError, AuthError} from './errors';

export class Defer<T = string> {
    resolve!: (text: T) => void;

    reject!: (error: any) => void;

    promise: Promise<T>;

    constructor() {
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
    }
}

export function bytes(texts: string[]) {
    return texts.reduce((sum, text) => sum + text.length, 0);
}
