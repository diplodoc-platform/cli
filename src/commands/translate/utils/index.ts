export type {Locale} from './config';
export {resolveSchemas, FileLoader} from './fs';
export {extract, compose} from './translate';
export {resolveSource, resolveTargets, resolveFiles, resolveVars} from './config';
export {
    TranslateError,
    ExtractError,
    ComposeError,
    SkipTranslation,
    EmptyTokensError,
} from './errors';
