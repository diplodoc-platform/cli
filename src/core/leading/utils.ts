import {cloneDeepWith, isString} from 'lodash';

export function modifyValuesByKeys(
    object: object,
    keys: string[],
    modify: (value: string) => string,
) {
    // Clone the object deeply with a customizer function that modifies matching keys
    return cloneDeepWith(object, (value: unknown, key) => {
        if (keys.includes(key as string) && isString(value)) {
            return modify(value);
        }

        return undefined;
    });
}
