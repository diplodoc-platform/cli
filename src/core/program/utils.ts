import type {BaseArgs, BaseConfig} from '~/core/program/types';
import type {BaseProgram} from '~/core/program/index';

import {isAbsolute} from 'node:path';

export function isRelative(path: string | undefined) {
    return /^\.{1,2}\//.test(path || '');
}

export function requireExtension(name: string) {
    if (isRelative(name) || isAbsolute(name)) {
        return require(name);
    }

    const local = (name: string) => {
        if (name.startsWith('@')) {
            return undefined;
        }

        return '@diplodoc/cli/' + name;
    };

    const scoped = (name: string, fixed = false) => {
        if (name.startsWith('@')) {
            return undefined;
        }

        return '@diplodoc/' + name + (fixed ? '-extension' : '');
    };

    const precise = (name: string, fixed = false) => {
        return name + (fixed ? '-extension' : '');
    };

    const variants = [
        local(name),
        scoped(name),
        scoped(name, true),
        precise(name),
        precise(name, true),
    ].filter(Boolean);

    while (variants.length) {
        const variant = variants.shift() as string;
        try {
            return require(variant);
        } catch {}
    }

    throw new Error(`Unable to resolve '${name}' extension`);
}

export function isProgram<TConfig extends BaseConfig, TArgs extends BaseArgs>(
    module: unknown,
): module is BaseProgram<TConfig, TArgs> {
    return Boolean(module && typeof (module as BaseProgram).init === 'function');
}
