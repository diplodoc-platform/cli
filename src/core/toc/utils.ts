import type {TextFilter} from './types';

import {evaluate} from '@diplodoc/liquid';

export function isRelative(path: AnyPath): path is RelativePath {
    return /^\.{1,2}\//.test(path) || !/^(\w{0,7}:)?\/\//.test(path);
}

export function getFirstValuable<T>(
    items: TextFilter[] | string,
    vars: Hash,
    fallback?: T,
): T | undefined {
    if (typeof items === 'string') {
        items = [{text: items, when: true}];
    }

    if (!Array.isArray(items)) {
        items = [];
    }

    for (const item of items) {
        let {when = true} = item;
        delete item.when;

        if (typeof when === 'string') {
            when = Boolean(evaluate(when, vars));
        }

        if (when) {
            return item.text as T;
        }
    }

    return fallback;
}
