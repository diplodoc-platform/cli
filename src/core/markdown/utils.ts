import type {AssetInfo, Location} from './types';

import {parseLocalUrl} from '~/core/utils';

type AssetModifier = '!' | '@' | '';

const modifiers = {'!': 'image', '@': 'video', '': 'link'} as const;

export function findLink(content: string): string | undefined {
    const rx = /]\(\s*/g;
    const match = rx.exec(content);
    if (!match) {
        return undefined;
    }

    const link = parseLinkDestination(content, rx.lastIndex);

    if (link === null || link.match(/[${}]/)) {
        return undefined;
    }

    rx.lastIndex += link.length + 1;

    return link;
}

export function findLinksInfo(content: string): AssetInfo[] {
    const links = find(/]\(\s*/g, content, (match, rx) => {
        const link = parseLinkDestination(content, rx.lastIndex);
        const title = parseLinkTitle(content, rx.lastIndex - match[0].length);

        // TODO: add more precise filter for unix compatible paths
        if (link === null || title === null || link.match(/[${}]/)) {
            return undefined;
        }

        const parsed = parseLocalUrl(link);
        if (!parsed) {
            return undefined;
        }

        const modifier = content[
            rx.lastIndex - match[0].length - title.length - 2
        ] as AssetModifier;
        const type = modifiers[modifier] || 'link';

        rx.lastIndex += link.length + 1;

        return {
            ...parsed,
            type,
            subtype: type === 'image' ? 'image' : null,
            title,
            autotitle: type === 'link' && (!title || title === '{#T}'),
        };
    });

    const referenceLinks = find(/]\[\s*/g, content, (match, rx) => {
        let link = parseLinkDestination(content, rx.lastIndex, ['[', ']']);
        let title = parseLinkTitle(content, rx.lastIndex - match[0].length);

        // TODO: add more precise filter for unix compatible paths
        if ((link === null && title === null) || link?.match(/[${}]/)) {
            return undefined;
        }
        title = title || '';
        link = link || title;

        const parsed = parseLocalUrl(link);
        if (!parsed) {
            return undefined;
        }

        const modifier = content[
            rx.lastIndex - match[0].length - title.length - 2
        ] as AssetModifier;
        const type = modifiers[modifier] || 'link';

        rx.lastIndex += link.length + 1;

        return {
            ...parsed,
            type,
            subtype: 'reference',
            code: link,
            title,
            autotitle: type === 'link' && (!title || title === '{#T}'),
        };
    });

    return [...links, ...referenceLinks];
}

export function findDefs(content: string): AssetInfo[] {
    return find(/^\s*\[(.*?)]:\s*([^\s]+)/gm, content, (match) => {
        const parsed = parseLocalUrl(match[2]);
        if (!parsed) {
            return undefined;
        }

        return {
            ...parsed,
            type: 'def',
            code: match[1] || undefined,
            title: '',
            autotitle: true,
        };
    });
}

function find(
    matcher: RegExp,
    content: string,
    map: (match: RegExpMatchArray, rx: RegExp) => Omit<AssetInfo, 'location'> | void,
): AssetInfo[] {
    const links: AssetInfo[] = [];

    let match;
    // eslint-disable-next-line no-cond-assign
    while ((match = matcher.exec(content))) {
        const result = map(match, matcher);
        if (!result) {
            continue;
        }

        if (result) {
            links.push({
                ...result,
                location: [match.index, matcher.lastIndex],
            });
        }
    }

    return links;
}

function parseLinkDestination(str: string, start: number, symbols = ['(', ')']) {
    const max = str.length;
    const [symbolStart, symbolEnd] = symbols.map((symbol) => symbol.charCodeAt(0));

    let code,
        level = 0,
        pos = start;

    if (str.charCodeAt(pos) === 0x3c /* < */) {
        return parseSimpleLink(str, start + 1);
    }

    while (pos < max) {
        code = str.charCodeAt(pos);

        if (code === 0x20) {
            break;
        }

        // ascii control characters
        if (code < 0x20 || code === 0x7f) {
            break;
        }

        if (code === 0x5c /* \ */ && pos + 1 < max) {
            if (str.charCodeAt(pos + 1) === 0x20) {
                break;
            }
            pos += 2;
            continue;
        }

        if (code === symbolStart /* ( or ] */) {
            level++;
            if (level > 32) {
                return null;
            }
        }

        if (code === symbolEnd /* ) or ] */) {
            if (level === 0) {
                break;
            }
            level--;
        }

        pos++;
    }

    if (start === pos) {
        return null;
    }
    if (level !== 0) {
        return null;
    }

    return str.slice(start, pos);
}

function parseLinkTitle(str: string, start: number) {
    let code,
        prev,
        level = 0,
        pos = start;

    while (pos >= 0) {
        code = str.charCodeAt(pos);
        prev = str.charCodeAt(pos - 1);

        // ascii control characters
        if (code < 0x20 || code === 0x7f) {
            break;
        }

        if (prev === 0x5c /* \ */) {
            pos -= 2;
            continue;
        }

        if (code === 93 /* ] */) {
            level++;
            if (level > 32) {
                return null;
            }
        }

        if (code === 91 /* [ */) {
            if (level === 1) {
                break;
            }
            level--;
        }

        pos--;
    }

    return str.slice(pos + 1, start);
}

function parseSimpleLink(str: string, start: number) {
    const max = str.length;

    let code,
        pos = start;

    while (pos < max) {
        code = str.charCodeAt(pos);
        if (code === 0x0a /* \n */) {
            return null;
        }
        if (code === 0x3c /* < */) {
            return null;
        }
        if (code === 0x3e /* > */) {
            return str.slice(start, pos);
        }
        if (code === 0x5c /* \ */ && pos + 1 < max) {
            pos += 2;
            continue;
        }

        pos++;
    }

    // no closing '>'
    return null;
}

export function filterRanges<T extends {location: Location}>(
    excludes: Location[],
    infos: T[],
): T[] {
    const contains = (exclude: Location, point: Location) => {
        return (
            (exclude[1] >= point[0] && exclude[1] <= point[1]) ||
            (exclude[0] >= point[0] && exclude[0] < point[1]) ||
            (exclude[0] < point[0] && exclude[1] >= point[1])
        );
    };

    return infos.filter((item) => {
        return !excludes.some((exclude) => contains(exclude, item.location));
    });
}

export function parseHeading(content: string) {
    const anchors = [];
    const commonHeading = content.match(/^#+/);
    const alternateHeading = content[content.length - 1];
    const alternaleLevels = ['-', '='];
    const level = commonHeading
        ? commonHeading[0].length
        : alternaleLevels.indexOf(alternateHeading) + 1;

    if (commonHeading) {
        content = content.replace(/^#+\s*/, '');
    } else {
        content = content.replace(/\n[-=]+$/, '');
    }

    const ANCHOR = /{(#[^}]+)}/g;

    let match;
    // eslint-disable-next-line no-cond-assign
    while ((match = ANCHOR.exec(content))) {
        anchors.push(match[1]);
        content = content.replace(match[0], '');
        ANCHOR.lastIndex -= match[0].length;
    }

    const title = content.trim();

    return {anchors, title, level};
}
