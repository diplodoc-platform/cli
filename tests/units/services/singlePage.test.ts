import {parse} from 'node-html-parser';
import {join} from 'path';
import {platformless} from '../../utils';

import {
    tryFixFirstPageHeader,
    addPagePrefixToAnchors,
    decreaseHeadingLevels,
    replaceLinks,
    replaceImages,
    addMainTitle,
} from 'utils/singlePage';

describe('Adding the main title', () => {
    const options = {
        title: 'Title',
        root: __dirname,
        path: 'about.html',
        tocDir: '',
    };

    test('Should be just added', () => {
        const content = '<h2></h2>';
        const root = parse(content);
        addMainTitle(root, options);
        expect(platformless(root.toString())).toMatchSnapshot();
    });

    test('Should be added the page attributes', () => {
        const content = '<h1>Title</h1>';
        const root = parse(content);
        addPagePrefixToAnchors(root, options);
        expect(platformless(root.toString())).toMatchSnapshot();
    });
});

describe('Try to fix the first page header', () => {
    test('Should be fixed', () => {
        const content = '<h2></h2><h2></h2>';
        const root = parse(content);
        tryFixFirstPageHeader(root);
        expect(platformless(root.toString())).toMatchSnapshot();
    });

    test('Should be ignored if all is correct', () => {
        const content = '<h1></h1><h2></h2>';
        const root = parse(content);
        tryFixFirstPageHeader(root);
        expect(platformless(root.toString())).toMatchSnapshot();
    });

    test('Should be ignored if there are no headers', () => {
        const content = '<div></div>';
        const root = parse(content);
        tryFixFirstPageHeader(root);
        expect(platformless(root.toString())).toMatchSnapshot();
    });
});

describe('Add a page prefix to anchors', () => {
    test('Should be added', () => {
        const content = [
            '<h1 id="first-header">',
            '<a class="yfm-anchor" href="#first-header" id="first-header"></a>',
            '</h1>',
            '<h2 id="second-header">',
            '<a class="yfm-anchor" href="#second-header" id="second-header"></a>',
            '</h2>',
        ].join('\n');

        const root = parse(content);
        addPagePrefixToAnchors(root, {
            root: __dirname,
            path: 'index.html',
            tocDir: '',
        });
        expect(platformless(root.toString())).toMatchSnapshot();
    });
});

describe('Decrease heading levels', () => {
    test('Should work', () => {
        const content = '<h1></h1><h1></h1><h2></h2><h3></h3><h4></h4><h5></h5><h6></h6>';
        const root = parse(content);
        decreaseHeadingLevels(root);
        expect(platformless(root.toString())).toMatchSnapshot();
    });
});

describe('Make links to pages as hashes to sections in a single page', () => {
    test('Should be replaced if file is root', () => {
        const content = [
            `<a href="folder/index.html"></a>`,
            `<a href="folder/index.html#third-header"></a>`,
            '<a href="index.html"></a>',
        ].join('\n');
        const root = parse(content);
        replaceLinks(root, {
            root: __dirname,
            path: 'index.html',
            tocDir: __dirname,
        });
        expect(platformless(root.toString())).toMatchSnapshot();
    });

    test('Should be replaced if file is not root', () => {
        const content = [
            `<a href="../folder/index.html"></a>`,
            `<a href="../folder/index.html#third-header"></a>`,
            `<a href="../index.html"></a>`,
        ].join('\n');
        const root = parse(content);
        replaceLinks(root, {
            root: __dirname,
            path: 'folder1/index.html',
            tocDir: __dirname,
        });
        expect(platformless(root.toString())).toMatchSnapshot();
    });

    test('External links should not be replaced', () => {
        const content = '<a href="https://ydocs.tech" target="_blank"></a>';
        const root = parse(content);
        replaceLinks(root, {
            root: __dirname,
            path: 'index.html',
            tocDir: __dirname,
        });
        expect(platformless(root.toString())).toMatchSnapshot();
    });

    test('Should not be replaced as hash if a link is out of the root folder', () => {
        const content = `<a href="../../index.html"></a>`;
        const root = parse(content);
        replaceLinks(root, {
            root: __dirname,
            path: 'folder/index.html',
            tocDir: __dirname,
        });
        expect(platformless(root.toString())).toMatchSnapshot();
    });

    test('Should not be replaced as hash if a link is out of the single page (but not out of the root)', () => {
        const content = `<a href="../../service2/index.html"></a>`;
        const root = parse(content);
        replaceLinks(root, {
            root: __dirname,
            path: 'service1/folder/index.html',
            tocDir: join(__dirname, 'service1'),
        });
        expect(platformless(root.toString())).toMatchSnapshot();
    });
});

describe('Make image sources relative from the single page file', () => {
    test('Image sources should not be replaced if file is root', () => {
        const content = [
            `<img src="_assets/index.png" alt="test">`,
            `<img src="../../_assets/index.png" alt="test">`,
        ].join('\n');
        const root = parse(content);
        replaceImages(root, {
            root: __dirname,
            path: 'index.html',
            tocDir: __dirname,
        });
        expect(platformless(root.toString())).toMatchSnapshot();
    });

    test('Image sources should be replaced if the file is not root', () => {
        const content = [
            `<img src="_assets/index.png" alt="test">`,
            `<img src="../../_assets/index.png" alt="test">`,
        ].join('\n');
        const root = parse(content);
        replaceImages(root, {
            root: __dirname,
            path: 'folder/index.html',
            tocDir: __dirname,
        });
        const result = root.toString();
        expect(platformless(result)).toMatchSnapshot();
    });

    test('External image sources should not be replaced', () => {
        const content = '<img src="https://ydocs.tech/favicon.png" alt="test">';
        const root = parse(content);
        replaceImages(root, {
            root: __dirname,
            path: 'index.html',
            tocDir: __dirname,
        });
        expect(platformless(root.toString())).toMatchSnapshot();
    });
});
