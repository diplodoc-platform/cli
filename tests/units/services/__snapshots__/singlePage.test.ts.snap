// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add a page prefix to anchors Should be added 1`] = `
<h1 id="_index_first-header"
    data-original-article="/index.html"
>
  <a class="yfm-anchor"
     href="#_index_first-header"
     id="_index_first-header"
  >
  </a>
  <a class="yfm-anchor"
     aria-hidden="true"
     href="#_index"
     id="_index"
  >
  </a>
</h1>
<h2 id="_index_second-header">
  <a class="yfm-anchor"
     href="#_index_second-header"
     id="_index_second-header"
  >
  </a>
</h2>
`;

exports[`Adding the main title Should be added the page attributes 1`] = `
<h1 data-original-article="/about.html">
  Title
  <a class="yfm-anchor"
     aria-hidden="true"
     href="#_about"
     id="_about"
  >
  </a>
</h1>
`;

exports[`Adding the main title Should be just added 1`] = `
<h1>
  Title
</h1>
<h2>
</h2>
`;

exports[`Decrease heading levels Should work 1`] = `
<h2>
</h2>
<h2>
</h2>
<h3>
</h3>
<h4>
</h4>
<h5>
</h5>
<h6>
</h6>
<h6>
</h6>
`;

exports[`Make image sources relative from the single page file External image sources should not be replaced 1`] = `
<img src="https://ydocs.tech/favicon.png"
     alt="test"
>
`;

exports[`Make image sources relative from the single page file Image sources should be replaced if the file is not root 1`] = `
<img src="folder/_assets/index.png"
     alt="test"
>
<img src="../_assets/index.png"
     alt="test"
>
`;

exports[`Make image sources relative from the single page file Image sources should not be replaced if file is root 1`] = `
<img src="_assets/index.png"
     alt="test"
>
<img src="../../_assets/index.png"
     alt="test"
>
`;

exports[`Make links to pages as hashes to sections in a single page External links should not be replaced 1`] = `
<a href="https://ydocs.tech"
   target="_blank"
>
</a>
`;

exports[`Make links to pages as hashes to sections in a single page Should be replaced if file is not root 1`] = `
<a href="#_folder_index">
</a>
<a href="#_folder_index_third-header">
</a>
<a href="#_index">
</a>
`;

exports[`Make links to pages as hashes to sections in a single page Should be replaced if file is root 1`] = `
<a href="#_folder_index">
</a>
<a href="#_folder_index_third-header">
</a>
<a href="#_index">
</a>
`;

exports[`Make links to pages as hashes to sections in a single page Should not be replaced as hash if a link is out of the root folder 1`] = `
<a href="../index.html">
</a>
`;

exports[`Make links to pages as hashes to sections in a single page Should not be replaced as hash if a link is out of the single page (but not out of the root) 1`] = `
<a href="../service2/index.html">
</a>
`;

exports[`Try to fix the first page header Should be fixed 1`] = `
<h1>
</h1>
<h2>
</h2>
`;

exports[`Try to fix the first page header Should be ignored if all is correct 1`] = `
<h1>
</h1>
<h2>
</h2>
`;

exports[`Try to fix the first page header Should be ignored if there are no headers 1`] = `
<div>
</div>
`;
